### ISSUE TITLE
Migrate to Symphonia to support more audio formats

### ISSUE THREAD

**Issue by mrDIMAS** [FyroxEngine/Fyrox#707]
A very common format and it should be supported by the sound system.

**Edit.** Symphonia looks to be a better replacement for current decoder implementation.

**Comment by pedroafabri** [comment]
I once worked with a sound engineer/composer that explained to me that mp3 are not very commom in games because it creates some miliseconds of silence at the start and end of the audio (he called it "silence tails" or something).

Since then I started using .ogg because it's as good as .wav but lighter and doesn't create those silence tails.

With that said, do you believe it's worth spending time supporting this format?

I mean... I use .mp3 for quick prototyping or game jams and the majority of game engines support it, but I decided to infodump here just in case lol

**Comment by pedroafabri** [comment]
@mrDIMAS which library would you sugest to decode the mp3 file?

I'm new to Rust but excited by it's universe and I found this library: https://github.com/germangb/minimp3-rs

Do you think it'll do the job?

**Comment by mrDIMAS** [comment]
@pedroafabri I think it is worth integrating [symphonia](https://crates.io/crates/symphonia) and remove all other decoder dependencies. It seems that this lib supports a ton of formats and this change could solve all the future format requests as well.

**Comment by mrDIMAS** [comment]
But for now you could add minimp3-rs if that's easier.

**Comment by pedroafabri** [comment]
@mrDIMAS maybe we could use symphonia for the .mp3 and open new issues to change .wav and .ogg to that lib too if it turns out well.

What do you think?

**Comment by mrDIMAS** [comment]
Symphonia looks to be [quite easy to use](https://github.com/pdeljanov/Symphonia/blob/master/symphonia/examples/getting-started.rs) and it could be a drop-in replacement for [Decoder](https://github.com/FyroxEngine/Fyrox/blob/master/fyrox-sound/src/decoder/mod.rs#L32). All that is needed is to basically use `next_packet` to read next bunch of samples and then use these samples in [Iterator](https://github.com/FyroxEngine/Fyrox/blob/master/fyrox-sound/src/decoder/mod.rs#L40) impl.

I understand your concern, this change is quite significant and could potentially be a source of problems, but `fyrox-sound` has a bunch of [examples](https://github.com/FyroxEngine/Fyrox/tree/master/fyrox-sound/examples) that can be used to test if everything works as it should. 

I can also rename this issue to `Use symphonia crate instead of separate audio decoders` so it better matches the actual task.

**Comment by pedroafabri** [comment]
Oh, I see now that you were proposing to use Symphonia alongside the other formats all in the same issue.

If you think it’s not too much for a single PR, you could update the title to reflect that this issue will switch everything to Symphonia and implement MP3 support.

Also, I’m relatively new to open-source collaboration, so feel free to correct me if needed!

**Comment by mrDIMAS** [comment]
Yeah, Symphonia is basically a full replacement for every decoder in `fyrox-sound` + it adds support for lots of new formats. It looks to be a relatively small PR - 3 changed files + 2 deleted (old decoders). So the idea of `fyrox-sound` decoder interface is simple - it must implement `Iterator` trait that emits samples one by one in interleaved format (`LRLRLR...`). Symphonia reads data by packets so you basically need to read a packet, then consume it in the iterator, when it depletes - read next packet and so on. Should be simple to implement. There are also a few additional methods like `time_seek`, `rewind`, etc. and there should also be no problems with it since Symphonia provides such methods too.

**Comment by pedroafabri** [comment]
Perfect! I will probably only be able to take a look at the first week of January.

If you're ok with that you can assign this one to me.

**Comment by pedroafabri** [comment]
@mrDIMAS I want your opinion on something.

I started working on this issue today and, while trying to run the examples, I noticed that none of the examples that read a file worked.

I then debugged and discovered that the full path must be passed to `DataSource::from_file`, but that's not happening on all the examples on `fyrox-sound`.

`fyrox-sound/examples/play_sound.rs:43`:

```rust
let door_open_buffer = SoundBufferResource::new_generic(
        fyrox_sound::futures::executor::block_on(DataSource::from_file(
            "examples/data/door_open.wav",
            // Load from the default resource io (File system)
            &FsResourceIo,
        ))
        .unwrap(),
    )
    .unwrap();
```

The above code produces an error when running `cargo run --example play_sound --release`, but works perfectly if the path passed to  `DataSource::from_file` is changed to `fyrox-sound/examples/data/door_open.wav`.

My question is: Can I update the examples to use the full path to the files, or am I missing something?

**Comment by mrDIMAS** [comment]
Just run `cargo run --example play_sound --release` from the `fyrox-sound` folder and it should solve the issue.

**Comment by pedroafabri** [comment]
I was about to say to assign this issue to somenone else because I'm having trouble finding time to do it.

You can check my WIP [HERE](https://github.com/pedroafabri/Fyrox/tree/feature/symphonia). I followed the same structure but used Symphonia to detect the file type, that would allow more extensions than just wav, mp3 and ogg.

**Comment by septante** [comment]
I would be happy to take this over if @mrDIMAS thinks the changes I made are reasonable - it just needs a little more polish (deciding on the error type for the result from new(), mostly) and then implementing it for other formats should be pretty straightforward.

**Comment by pedroafabri** [comment]
If both of you are OK with it I don't mind.
Feel free to assign this to @septante, @mrDIMAS !

**Comment by mrDIMAS** [comment]
Yeah, sure. I'm fine with pretty much any reasonable solutions that _just works_.
